httpx/_config.py

```diff
@@ -42,150 +40,102 @@ class UnsetType:
-        trust_env: bool = True,
-        http2: bool = False,
-    ) -> None:
-        self.cert = cert
-        self.verify = verify
-        self.trust_env = trust_env
-        self.http2 = http2
-        self.ssl_context = self.load_ssl_context()
+        cert: typing.Optional[CertTypes] = None,
+    ) -> "SSLContext":
+        self = super().__new__(cls, protocol)
+
+        set_minimum_tls_version_1_2(self)
+        self.options |= ssl.OP_NO_COMPRESSION
+        self.set_ciphers(DEFAULT_CIPHERS)
 
-    def load_ssl_context(self) -> ssl.SSLContext:
         logger.debug(
-            "load_ssl_context verify=%r cert=%r trust_env=%r http2=%r",
-            self.verify,
-            self.cert,
-            self.trust_env,
-            self.http2,
+            "load_ssl_context verify=%r cert=%r",
+            verify,
+            cert,
         )
 
-        if self.verify:
-            return self.load_ssl_context_verify()
-        return self.load_ssl_context_no_verify()
+        if verify:
+            self.load_ssl_context_verify(cert, verify)
+        else:
+            self.load_ssl_context_no_verify(cert)
+        return self
 
-    def load_ssl_context_no_verify(self) -> ssl.SSLContext:
+    def load_ssl_context_no_verify(
+        self, cert: typing.Optional[CertTypes]
+    ) -> ssl.SSLContext:
         """
         Return an SSL context for unverified connections.
         """
-        context = self._create_default_ssl_context()
-        context.check_hostname = False
-        context.verify_mode = ssl.CERT_NONE
-        self._load_client_certs(context)
-        return context
+        self.check_hostname = False
+        self.verify_mode = ssl.CERT_NONE
+        self._load_client_certs(cert)
+        return self
 
-    def load_ssl_context_verify(self) -> ssl.SSLContext:
+    def load_ssl_context_verify(
+        self, cert: typing.Optional[CertTypes], verify: VerifyTypes
+    ) -> None:
         """
         Return an SSL context for verified connections.
         """
-        if self.trust_env and self.verify is True:
-            ca_bundle = get_ca_bundle_from_env()
-            if ca_bundle is not None:
-                self.verify = ca_bundle
-
-        if isinstance(self.verify, ssl.SSLContext):
-            # Allow passing in our own SSLContext object that's pre-configured.
-            context = self.verify
-            self._load_client_certs(context)
-            return context
-        elif isinstance(self.verify, bool):
+        if isinstance(verify, bool):
             ca_bundle_path = self.DEFAULT_CA_BUNDLE_PATH
-        elif Path(self.verify).exists():
-            ca_bundle_path = Path(self.verify)
+        elif Path(verify).exists():
+            ca_bundle_path = Path(verify)
         else:
             raise IOError(
                 "Could not find a suitable TLS CA certificate bundle, "
-                "invalid path: {}".format(self.verify)
+                "invalid path: {}".format(verify)
             )
```

Comment:
We confuse the flow here by having this section of code separated from where `ca_bundle_path` is actually used. It'd be more clear if we pushed this block down to just before `if ca_bundle_path.is_file():`.

Reply:
(Tho perhaps we address clean-ups separately to the main API behavioral changes here?)

Reply:
Separating them will make it easier to review this PR, which contains an important API change.

Reply:
Another important question is whether we need to deprecate previous arguments or simply delete them.

Reply:
> (Tho perhaps we address clean-ups separately to the main API behavioral changes here?)

Agree with tom, you could add `ssl_context=` and `httpx.SSLContext` in current PR and decide to deprecate/remove in separated PR.